// react-svg-pan-zoom/build-commonjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TOOL_ZOOM_OUT = exports.TOOL_ZOOM_IN = exports.TOOL_PAN = exports.TOOL_NONE = exports.TOOL_AUTO = exports.POSITION_TOP = exports.POSITION_RIGHT = exports.POSITION_NONE = exports.POSITION_LEFT = exports.POSITION_BOTTOM = exports.MODE_ZOOMING = exports.MODE_PANNING = exports.MODE_IDLE = exports.INITIAL_VALUE = exports.ALIGN_TOP = exports.ALIGN_RIGHT = exports.ALIGN_LEFT = exports.ALIGN_COVER = exports.ALIGN_CENTER = exports.ALIGN_BOTTOM = exports.ACTION_ZOOM = exports.ACTION_PAN = void 0;
var MODE_IDLE = 'idle';
exports.MODE_IDLE = MODE_IDLE;
var MODE_PANNING = 'panning';
exports.MODE_PANNING = MODE_PANNING;
var MODE_ZOOMING = 'zooming';
exports.MODE_ZOOMING = MODE_ZOOMING;
var TOOL_AUTO = 'auto';
exports.TOOL_AUTO = TOOL_AUTO;
var TOOL_NONE = 'none';
exports.TOOL_NONE = TOOL_NONE;
var TOOL_PAN = 'pan';
exports.TOOL_PAN = TOOL_PAN;
var TOOL_ZOOM_IN = 'zoom-in';
exports.TOOL_ZOOM_IN = TOOL_ZOOM_IN;
var TOOL_ZOOM_OUT = 'zoom-out';
exports.TOOL_ZOOM_OUT = TOOL_ZOOM_OUT;
var POSITION_NONE = 'none';
exports.POSITION_NONE = POSITION_NONE;
var POSITION_TOP = 'top';
exports.POSITION_TOP = POSITION_TOP;
var POSITION_RIGHT = 'right';
exports.POSITION_RIGHT = POSITION_RIGHT;
var POSITION_BOTTOM = 'bottom';
exports.POSITION_BOTTOM = POSITION_BOTTOM;
var POSITION_LEFT = 'left';
exports.POSITION_LEFT = POSITION_LEFT;
var ACTION_ZOOM = 'zoom';
exports.ACTION_ZOOM = ACTION_ZOOM;
var ACTION_PAN = 'pan';
exports.ACTION_PAN = ACTION_PAN;
var ALIGN_CENTER = 'center';
exports.ALIGN_CENTER = ALIGN_CENTER;
var ALIGN_LEFT = 'left';
exports.ALIGN_LEFT = ALIGN_LEFT;
var ALIGN_RIGHT = 'right';
exports.ALIGN_RIGHT = ALIGN_RIGHT;
var ALIGN_TOP = 'top';
exports.ALIGN_TOP = ALIGN_TOP;
var ALIGN_BOTTOM = 'bottom';
exports.ALIGN_BOTTOM = ALIGN_BOTTOM;
var ALIGN_COVER = 'cover';
exports.ALIGN_COVER = ALIGN_COVER;
var INITIAL_VALUE = {};
exports.INITIAL_VALUE = INITIAL_VALUE;
